rlw/upgrade.ps1

```diff
@@ -0,0 +1,167 @@
+      Stop-Process -Name 'rightlink' -Force
+    } else {
+      Write-Output """Error: ${upgradeCheck}"""
+      Exit 1
+    }
+    # Check updated version in production by connecting to local proxy
+    # The update takes a few seconds so retries are done.
+    for ($i = 1; $i -le 5; $i += 1) {
+      $newInstalledVersion = & ${RIGHTLINK_DIR}\rsc.exe --x1 .version rl10 index proc 2> $null
+      if ($newInstalledVersion -eq $desiredVersion) {
+        Write-Output """New version active - ${newInstalledVersion}"""
+        break
+      } else {
+        Write-Output 'Waiting for new version to become active.'
+        Sleep 2
+      }
+    }
+    if ($newInstalledVersion -ne $desiredVersion) {
+      Write-Output """New version does not appear to be desired version: ${newInstalledVersion}"""
+      # Put the old version of RightLink back because something seemed to have gone wrong
+      Rename-Item -Path ${RIGHTLINK_DIR}\rightlink.exe -newName 'rightlink-new.exe' -Force
+      Rename-Item -Path ${RIGHTLINK_DIR}\rightlink-old.exe -newName 'rightlink.exe' -Force
+      # Stop the new rightlink process so that we revert back to the old version when NSSM restarts RightLink
+      Stop-Process -Name 'rightlink' -Force
+      Exit 1
+    }
+
+    # Report to audit entry that RightLink ugpraded.
+    $instanceHref = & ${RIGHTLINK_DIR}\rsc.exe --rl10 --x1 ':has(.rel:val(\"\"\"self\"\"\")).href' `
+                    cm15 index_instance_session /api/sessions/instance 2> $null
+    if ($instanceHref) {
+      $auditEntryHref = & ${RIGHTLINK_DIR}\rsc.exe --rl10 --xh 'location' cm15 create /api/audit_entries `
+                        """audit_entry[auditee_href]=${instanceHref}""" `
+                        """audit_entry[detail]=RightLink updated to '${newInstalledVersion}'""" `
+                        """audit_entry[summary]=RightLink updated""" 2> $null
+      if ($auditEntryHref) {
+        Write-Output """Audit entry created at ${auditEntryHref}"""
+      } else {
+        Write-Output 'Failed to create audit entry'
+      }
+    } else {
+      Write-Output 'Unable to obtain instance HREF for audit entries'
+    }
+  }
+}
+
+$RIGHTLINK_DIR = 'C:\Program Files\RightScale\RightLink'
+
+if (!$env:UPGRADES_FILE_LOCATION) {
+  $env:UPGRADES_FILE_LOCATION = 'https://rightlink.rightscale.com/rightlink/upgrades'
+}
+
+# Query RightLink info
+$json = & "${RIGHTLINK_DIR}\rsc.exe" rl10 index /rll/proc
+
+# Determine current version of rightlink
+$currentVersion = Write-Output $json | & "${RIGHTLINK_DIR}\rsc.exe" --x1 .version json 2> $null
+
+if (!$currentVersion) {
+  Write-Output 'Cannot determine current version of RightLink'
+  Exit 1
+}
+
+# Fetch information about what we should become. The "upgrades" file consists of lines formatted
+# as "currentVersion:desiredVersion". If the "upgrades" file does not exist,
+# or if the current version is not in the file, no upgrade is done.
+$upgradesFile = "${PSScriptRoot}\upgrades"
```

Comment:
Do we really need to save this in a file? Would it be okay to just do `DownloadString` instead?